gateways/firstdata/firstdata.go

```diff
@@ -0,0 +1,215 @@
+		Response:             string(firstdataResponse.TransactionState),
+		AvsResultRaw:         avsRawString,
+		CvvResultRaw:         string(firstdataResponse.Processor.SecurityCodeResponse),
+	}, nil
+}
+
+func (client *FirstdataClient) Capture(request *sleet.CaptureRequest) (*sleet.CaptureResponse, error) {
+	firstdataCaptureRequest, err := buildCaptureRequest(request)
+	if err != nil {
+		return nil, err
+	}
+
+	firstdataResponse, err := client.sendRequest(
+		*request.ClientTransactionReference,
+		client.secondaryURL(request.TransactionReference),
+		*firstdataCaptureRequest,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if firstdataResponse.Error != nil {
+		response := sleet.CaptureResponse{Success: false, ErrorCode: &firstdataResponse.Error.Code}
+		return &response, nil
+	}
+
+	return &sleet.CaptureResponse{Success: true, TransactionReference: firstdataResponse.IPGTransactionId}, nil
+}
+
+// Void cancels a Firstdata payment. If successful, the void response will be returned. A previously voided
+// payment or one that has already been settled cannot be voided.
+func (client *FirstdataClient) Void(request *sleet.VoidRequest) (*sleet.VoidResponse, error) {
+	firstdataVoidRequest, err := buildVoidRequest(request)
+	if err != nil {
+		return nil, err
+	}
+
+	firstdataResponse, err := client.sendRequest(
+		*request.ClientTransactionReference,
+		client.secondaryURL(request.TransactionReference),
+		*firstdataVoidRequest,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if firstdataResponse.Error != nil {
+		response := sleet.VoidResponse{Success: false, ErrorCode: &firstdataResponse.Error.Code}
+		return &response, nil
+	}
+	return &sleet.VoidResponse{Success: true, TransactionReference: firstdataResponse.IPGTransactionId}, nil
+}
+
+// Refund refunds a Firstdata payment.
+// Multiple refunds can be made on the same payment, but the total amount refunded should not exceed the payment total.
+func (client *FirstdataClient) Refund(request *sleet.RefundRequest) (*sleet.RefundResponse, error) {
+	firstdataRefundRequest, err := buildRefundRequest(request)
+	if err != nil {
+		return nil, err
+	}
+
+	firstdataResponse, err := client.sendRequest(
+		*request.ClientTransactionReference,
+		client.secondaryURL(request.TransactionReference),
+		*firstdataRefundRequest,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if firstdataResponse.Error != nil {
+		response := sleet.RefundResponse{Success: false, ErrorCode: &firstdataResponse.Error.Code}
+		return &response, nil
+	}
+	return &sleet.RefundResponse{Success: true, TransactionReference: firstdataResponse.IPGTransactionId}, nil
+}
+
+func (client *FirstdataClient) sendRequest(reqId, url string, data Request) (*Response, error) {
+
+	bodyJSON, err := json.Marshal(data)
+	if err != nil {
+		return nil, err
+	}
+
+	timestamp := strconv.FormatInt(time.Now().Unix(), 10)
+
+	hashData := client.apiKey + reqId + timestamp + string(bodyJSON)
```

Comment:
can we paste a URL showing the requirements for singing the body

Reply:
Place the link as a comment in the code or just here ?

Reply:
Paste a comment as the link in code